sys/drv/hidkbd.c

```diff
@@ -0,0 +1,212 @@
+/*
+ * USB HID keyboard driver.
+ *
+ * Relevant documents:
+ *
+ * - Device Class Definition for Human Interface Devices (HID) Version 1.11,
+ *   5/27/01:
+ *     https://www.usb.org/sites/default/files/hid1_11.pdf
+ */
+#define KL_LOG KL_DEV
+#include <sys/klog.h>
+#include <sys/libkern.h>
+#include <sys/errno.h>
+#include <sys/devclass.h>
+#include <sys/device.h>
+#include <sys/sched.h>
+#include <dev/evdev.h>
+#include <dev/usb.h>
+
+#define HIDKBD_NMODKEYS 8
+#define HIDKBD_NKEYCODES 6
+
+/* HID keyboard input report. */
+typedef struct hidkbd_in_report {
+  uint8_t modifier_keys;
+  uint8_t reserved;
+  uint8_t keycodes[HIDKBD_NKEYCODES];
+} __packed hidkbd_in_report_t;
+
+/* HID keyboard software context. */
+typedef struct hidkbd_state {
+  hidkbd_in_report_t prev_report; /* the most recent report */
+  thread_t *thread;               /* thread which gathers scancodes */
+  evdev_dev_t *evdev;             /* evdev device structure */
+} hidkbd_state_t;
+
+/*
+ * Evdev handling functions.
+ */
+
+/* Create and setup a corresponding evdev device. */
+static void hidkbd_init_evdev(device_t *dev) {
+  usb_device_t *udev = usb_device_of(dev);
+  hidkbd_state_t *hidkbd = dev->state;
+  evdev_dev_t *evdev = evdev_alloc();
+
+  /* Set the basic evdev parameters. */
+  evdev_set_name(evdev, "HID keyboard");
+  evdev_set_id(evdev, BUS_USB, udev->vendor_id, udev->product_id, 0);
+
+  /* EV_SYN is required for every evdev device. */
+  evdev_support_event(evdev, EV_SYN);
+
+  evdev_support_event(evdev, EV_KEY);
+  evdev_hidkbd_support_all_known_keys(evdev);
+  /* NOTE: we assume that the keyboard supports hardware key repetiton,
+   * which isn't said to always be true. */
+  evdev_support_event(evdev, EV_REP);
+
+  /* Bind `evdev` to `hidkbd` for future references. */
+  hidkbd->evdev = evdev;
+
+  /* Finally, register the device in the file system. */
+  evdev_register(evdev);
+}
+
+/*
+ * Report processing functions.
+ */
+
+/* Process modifier keys. */
+static void hidkbd_process_modkeys(hidkbd_state_t *hidkbd, uint8_t modkeys) {
+  uint8_t prev_modkeys = hidkbd->prev_report.modifier_keys;
+
+  for (size_t i = 0; i <= HIDKBD_NMODKEYS; i++) {
+    uint8_t prev = prev_modkeys & (1 << i);
+    uint8_t cur = modkeys & (1 << i);
+
+    if (prev == cur)
+      continue;
+
+    uint16_t keycode = evdev_mod2key(i);
+    int32_t value = cur ? 1 : 0;
+    evdev_push_event(hidkbd->evdev, EV_KEY, keycode, value);
```

Comment:
There's a wrapper for this function. It could be just `evdev_push_key(hidkbd->evdev, keycode, value);`. Now I have noticed that I did the same in the `atkbdc` driver, so it could be changed as well.

Reply:
Done!